File: test/predictions/prediction_history_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:sreadya/features/predictions/domain/cycle_prediction.dart';
import 'package:sreadya/features/predictions/domain/prediction_history.dart';

CyclePrediction prediction({required PredictionConfidence confidence}) =>
    CyclePrediction(
      algorithmVersion: 'prediction-v1',
      mostLikelyDate: DateTime(2026, 9, 20),
      windowStart: DateTime(2026, 9, 18),
      windowEnd: DateTime(2026, 9, 22),
      estimatedCycleLengthDays: 29,
      confidence: confidence,
      validIntervals: const [29, 29, 28, 30, 29, 29],
      excludedIntervals: const [],
      medianAbsoluteDeviation: 1,
      createdAt: DateTime(2026, 9, 1),
    );

void main() {
  test('evaluates absolute error and prediction-window coverage', () {
    final result = PredictionEvaluator().evaluate([
      PredictionOutcome(
        prediction: prediction(confidence: PredictionConfidence.high),
        actualStart: DateTime(2026, 9, 21),
      ),
    ]);

    expect(result.meanAbsoluteErrorDays, 1);
    expect(result.medianAbsoluteErrorDays, 1);
    expect(result.windowCoverage, 1);
    expect(result.lateBiasDays, 1);
    expect(result.overConfidenceRate, 0);
  });

  test('reports over-confidence and under-confidence calibration signals', () {
    final result = PredictionEvaluator().evaluate([
      PredictionOutcome(
        prediction: prediction(confidence: PredictionConfidence.high),
        actualStart: DateTime(2026, 9, 27),
      ),
      PredictionOutcome(
        prediction: prediction(confidence: PredictionConfidence.low),
        actualStart: DateTime(2026, 9, 20),
      ),
    ]);

    expect(result.overConfidenceRate, 1);
    expect(result.underConfidenceRate, 1);
  });
}

File: web/src/domain/prediction/prediction-history.test.ts
import { describe, expect, it } from 'vitest';

import type { PredictionResult } from '../../features/predictions/prediction-engine';
import { evaluatePredictionHistory } from './prediction-history';

function prediction(
  confidence: PredictionResult['confidence'],
  mostLikelyDate: string,
  windowStart: string,
  windowEnd: string,
): PredictionResult {
  return {
    algorithmVersion: 'prediction-v1',
    estimatedCycleLengthDays: 28,
    estimatedPeriodDurationDays: 5,
    mostLikelyDate,
    windowStart,
    windowEnd,
    confidence,
    validIntervals: [28, 28, 28],
    excludedIntervals: [],
    medianAbsoluteDeviation: 0,
  };
}

describe('prediction-v1 history evaluation', () => {
  it('matches the established mobile accuracy semantics', () => {
    const evaluation = evaluatePredictionHistory([
      {
        prediction: prediction('high', '2026-04-23', '2026-04-21', '2026-04-25'),
        actualStart: '2026-04-24',
      },
      {
        prediction: prediction('low', '2026-05-20', '2026-05-18', '2026-05-22'),
        actualStart: '2026-05-20',
      },
    ]);

    expect(evaluation).toEqual({
      sampleCount: 2,
      meanAbsoluteErrorDays: 0.5,
      medianAbsoluteErrorDays: 0.5,
      windowCoverage: 1,
      earlyBiasDays: 0,
      lateBiasDays: 1,
      overConfidenceRate: 0,
      underConfidenceRate: 1,
    });
  });
});
